keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,487 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, a Kubernetes Service accepts a list of ports to be exposed by it.
+It is possible to specify any number of ports (as long as the service is within the max object size limit), by listing them in the service spec.
+This can be tedious if the service needs a large number of ports.
+This KEP proposes to add a new field to the Service spec to allow exposing the entire port range(1 to 65535).
+
+## Motivation
+
+There are several applications like SIP apps or RTP which needs a lot of ports to run multiple calls or media streams.
+Currently the only option is to specify every port in the Service Spec. A request for port ranges in Services has been open in https://github.com/kubernetes/kubernetes/issues/23864. Implementing port ranges are challenging since iptables/ipvs do not support remapping port ranges. Hence, the proposal to set a single field in order to expose the entire port range and implement the service clients and endpoints accordingly.
+[A survey](https://docs.google.com/forms/d/1FOOG2ZoQsnJLYAjnhEtSPYmUULWFNe88iXR7gtFcP7g/edit) was sent out to collect the use-cases for AllPorts support - [results.](http://tiny.cc/allportsslides)
+
+### Goals
+
+* Allow users to optionally expose the entire port range via a Service (of Type LoadBalancer or ClusterIP).
+
+### Non-Goals
+
+* Supporting Port Ranges in a Service.
+* Changing the default behavior of Service ports.
+
+## Proposal
+
+The proposal here is to introduce an `allPorts` boolean field to the [service API.](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#servicespec-v1-core)
```

Comment:
The new field has to be `*bool` not `bool`.  Also it would make the KEP alot relatable if you can add
1. sample service.
2. upgrade downgrade rules. e.g. what will happen to existing services if the user upgraded to newer version with this feature enabled? .. similarly what will happen to services when the feature is turned of?
3. The behavior re conversions. Services are have some weirder behavior (than most other api objects) . Example headful->headless, ClusterIP->LoadBalancer. etc. The list will help minimizing surprises during development of code.

Reply:
Done.